### ISSUE TITLE
Patch 8.2.1297 breaks s:function call for sourcing kana/vim-arpeggio setup

### ISSUE THREAD

**Issue by atweiden** [vim/vim#6736]
**Describe the bug**

Bisected to patch 8.2.1297. Seemingly throws the `<SID>work` function in https://github.com/kana/vim-arpeggio/raw/master/autoload/arpeggio.vim into an infinite loop.

**To Reproduce**

```bash
git clone https://github.com/kana/vim-arpeggio ~/.vim/pack/keyboard/opt/vim-arpeggio

cat > vimrc <<'EOF'
set nocp
function! s:source_cfg()
  source arpeggio.vim
endfunction
call s:source_cfg()
EOF

cat > arpeggio.vim <<'EOF'
packadd vim-arpeggio
Arpeggio inoremap jk <ESC>
EOF

vim -u vimrc
```

```vim
" enter insert mode
" leave insert mode by chording 'j' and 'k'
" vim freezes: Ctrl+C
```

Fix:

```bash
cat > vimrc <<'EOF'
set nocp
source arpeggio.vim
EOF
```

The `call s:source_cfg()` makes all the difference.

**Expected behavior**

[kana/vim-arpeggio](https://github.com/kana/vim-arpeggio) escape chord exits insert mode.

**Screenshots**

![](https://i.imgur.com/Aoo8YnB.png)

**Environment (please complete the following information):**

 - Vim version: 8.2.1297+
 - OS: macOS 10.15.6
 - Terminal: iTerm2

**Comment by brammool** [comment]
Can you dig up more information about where something goes wrong?  E.g. is the problem with the s:SID variable?

**Comment by atweiden** [comment]
Fixed pending https://github.com/kana/vim-arpeggio/pull/31.

**Comment by lacygoill** [comment]
> Can you dig up more information about where something goes wrong? E.g. is the problem with the s:SID variable?

I don't know whether it's a bug, but I can reduce the issue to this:

    vim -Nu NONE -S <(cat <<'EOF'
        vim9script
        let lines =<< trim END
            vim9script
            def Rhs(): string
                return s:SID .. 'j'
            enddef
            def SID(): string
                return expand('<sfile>')->matchstr('<SNR>\zs\d\+\ze_')
            enddef
            let SID = "\<SNR>" .. s:SID() .. '_'
            imap <expr> j <sid>Rhs()
            ino <sid>j j
        END
        writefile(lines, '/tmp/t.vim')
        def Source()
          so /tmp/t.vim
        enddef
        Source()
    EOF
    )

It doesn't require any plugin.

On 8.2.1296, if you press `j` in insert mode, Vim just inserts `j`.
On 8.2.1297, if you press `j` in insert mode, Vim expands the recursive mapping indefinitely; this one:

    imap <expr> j <sid>Rhs()

**Comment by lacygoill** [comment]
> Fixed pending kana/vim-arpeggio#31.

Even if the current issue is fixed, your code will still not work as expected, because of another bug in the arpeggio plugin:
https://github.com/kana/vim-arpeggio/blob/5d73721898d2bb669ead417b4d3370921ce7c3cf/autoload/arpeggio.vim#L267

It assumes that the first script ID in the stack of function calls is the one where `s:SID()` is defined; that's not necessarily true.  The last script ID is the one where `s:SID` is defined; not the first.

Here is a patch which should fix the bug:
```vim
diff --git a/autoload/arpeggio.vim b/autoload/arpeggio.vim
index 3e14315..68c26aa 100644
--- a/autoload/arpeggio.vim
+++ b/autoload/arpeggio.vim
@@ -264,7 +264,7 @@ endfunction
 
 " Misc.  "{{{1
 function! s:SID()  "{{{2
-  return matchstr(expand('<sfile>'), '<SNR>\zs\d\+\ze_')
+  return matchstr(expand('<sfile>'), '<SNR>\zs\d\+\ze_SID$')
 endfunction
 let s:SID = "\<SNR>" . s:SID() . '_'
```
